exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -289,23 +292,58 @@ private void checkTransaction(TransactionMessage transactionMessage) {
 
   /**
    * Performs a given function with a snapshot of the current database state (i.e., the one that
-   * corresponds to the latest committed block). In-pool (not yet processed) transactions are also
-   * accessible with it in {@linkplain Blockchain#getTxMessages() blockchain}.
+   * corresponds to the latest committed block) and returns a result of its execution. In-pool
+   * (not yet processed) transactions are also accessible with it in
+   * {@linkplain Blockchain#getTxMessages() blockchain}.
    *
    * @param snapshotFunction a function to execute
    * @param <ResultT> a type the function returns
    * @return the result of applying the given function to the database state
    */
-  public <ResultT> ResultT withSnapshot(Function<Snapshot, ResultT> snapshotFunction) {
-    try (Cleaner cleaner = new Cleaner("TestKit#withSnapshot")) {
-      long snapshotHandle = nativeCreateSnapshot(nativeHandle.get());
-      Snapshot snapshot = Snapshot.newInstance(snapshotHandle, cleaner);
+  public <ResultT> ResultT applySnapshot(Function<Snapshot, ResultT> snapshotFunction) {
+    try (Cleaner cleaner = new Cleaner("TestKit#applySnapshot")) {
+      Snapshot snapshot = createSnapshot(cleaner);
       return snapshotFunction.apply(snapshot);
     } catch (CloseFailuresException e) {
       throw new RuntimeException(e);
     }
   }
 
+  /**
+   * Performs a given function with a snapshot of the current database state (i.e., the one that
+   * corresponds to the latest committed block). In-pool (not yet processed) transactions are also
+   * accessible with it in {@linkplain Blockchain#getTxMessages() blockchain}.
+   *
+   * @param snapshotFunction a function to execute
+   */
+  public void withSnapshot(Consumer<Snapshot> snapshotFunction) {
+    try (Cleaner cleaner = new Cleaner("TestKit#withSnapshot")) {
+      Snapshot snapshot = createSnapshot(cleaner);
+      snapshotFunction.accept(snapshot);
+    } catch (CloseFailuresException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Returns a snapshot of the current database state (i.e., the one that
+   * corresponds to the latest committed block). In-pool (not yet processed) transactions are also
+   * accessible with it in {@linkplain Blockchain#getTxMessages() blockchain}.
+   *
+   * <p>All created snapshots are deleted when corresponding TestKit is disposed.
```

Comment:
```java
   * <p>All created snapshots are deleted when this TestKit is {@linkplain #close() closed}.
   * It is forbidden to access the snapshots once the TestKit is closed.
   *
   * <p>If you create many snapshots (more than a hundred), it is recommended
   * to use {@link #withSnapshot(Consumer)} or {@link #applySnapshot(Function)},
   * which destroy the snapshots once the passed closure completes.
```

Reply:
Not sure about a hundred, actually, any number will not _generally_ work because you can create five, but each will allocate a million indexes. Explaining that, however, would require explaining how our proxies and cleaners work — but their point is to isolate the users from this knowledge.